lib/LanguageServerRename/Model/Renamer/MemberRenamer.php

```diff
@@ -0,0 +1,182 @@
+    private $parser;
+    /** @var NodeUtils */
+    private $nodeUtils;
+    /** @var TextDocumentLocator */
+    private $locator;
+    /** @var RenameLocationsProvider */
+    private $locationsProvider;
+
+    public function __construct(
+        Parser $parser,
+        NodeUtils $nodeUtils,
+        TextDocumentLocator $locator,
+        RenameLocationsProvider $locationsProvider
+    ) {
+        $this->parser = $parser;
+        $this->nodeUtils = $nodeUtils;
+        $this->locator = $locator;
+        $this->locationsProvider = $locationsProvider;
+    }
+
+    private function getValidNode(TextDocument $textDocument, ByteOffset $offset): ?Node
+    {
+        $rootNode = $this->parser->parseSourceFile((string)$textDocument);
+        $node = $rootNode->getDescendantNodeAtPosition($offset->toInt());
+        
+        if (
+            $node instanceof MethodDeclaration
+            || ($node instanceof ScopedPropertyAccessExpression && $node->memberName instanceof Token)
+            || ($node instanceof Variable && $node->getFirstAncestor(PropertyDeclaration::class))
+            || ($node instanceof Variable && $node->getFirstAncestor(ScopedPropertyAccessExpression::class))
+            || $node instanceof ConstElement
+            || $node instanceof ClassConstDeclaration
+            || ($node instanceof MemberAccessExpression && $node->memberName instanceof Token)
+        ) {
+            return $node;
+        }
+        
+        return null;
+    }
+
+    public function prepareRename(TextDocument $textDocument, ByteOffset $offset): ?ByteOffsetRange
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) !== null) {
+            [ $token ] = $this->getNodeNameTokens($node);
+            return $this->nodeUtils->getTokenNameRange($token, (string)$textDocument);
+        }
+        
+        return null;
+    }
+    /**
+     * {@inheritDoc}
+     */
+    public function rename(TextDocument $textDocument, ByteOffset $offset, string $newName): Generator
+    {
+        if (($node = $this->getValidNode($textDocument, $offset)) === null) {
+            return null;
+        }
+
+        [ $token ] = $this->getNodeNameTokens($node);
+        $oldName = $this->nodeUtils->getTokenNameText($token, (string)$textDocument);
+
+        foreach ($this->locationsProvider->provideLocations($textDocument, $offset) as $locationGroup) {
+            /** @var RenameLocationGroup $locationGroup */
+            yield new RenameResult(
+                $this->locationsToTextEdits($locationGroup->uri(), $locationGroup->locations(), $oldName, $newName),
+                $locationGroup->uri()
+            );
+        }
+    }
+    /** @param Location[] $locations */
+    private function locationsToTextEdits(TextDocumentUri $textDocumentUri, array $locations, string $oldName, string $newName): TextEdits
+    {
+        $textEdits = [];
+        $rootNode = $this->parser->parseSourceFile($this->locator->get($textDocumentUri));
```

Comment:
Note that another strategy would be to not use the locations provided by the reference finder - references from the index can be out-of-date (the watcher could have stopped, it's too busy updating etc).

For the task of renaming members etc it's probably better to just find members with matching names in the given document.,

Reply:
But members could have been used in other documents?

Reply:
yes absolutely, we would still use the URL - it just seems safer to traverse the node tree rather than relying on the locations. Is there any advantage to using the potentially out-of-date locations?

Reply:
Yes, because members need to be resolved, which means invoking worse reflection. There could be a ton of objects in the file using the same name for a method which is not really the same method. I actually think it is out of the question to not use the index. I've been using a version of this code since Christmas and I had bad index only one time. 99% of the time it works perfectly. 

We could probably use the documents from the locations and re-parse all the documents, but that is scary as well.

Reply:
ok then :+1:

Reply:
as in "keep using the index"?

Reply:
yes